tsdb/agent/db_test.go

```diff
@@ -934,6 +935,177 @@ func TestDBOutOfOrderTimeWindow(t *testing.T) {
+			expectedType:        "histogram",
+			expectedErrorMsg:    storage.ErrCTNewerThanSample.Error(),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			reg := prometheus.NewRegistry()
+			opts := DefaultOptions()
+			opts.OutOfOrderTimeWindow = 0
+			s := createTestAgentDB(t, reg, opts)
+			app := s.Appender(context.TODO())
+
+			seriesMap := make(map[storage.SeriesRef]labels.Labels)
+			samplesMap := make(map[storage.SeriesRef][]record.RefSample)
+			histogramsMap := make(map[storage.SeriesRef][]record.RefHistogramSample)
+
+			lbls := labelsForTest(t.Name(), 1)
+			lset := labels.New(lbls[0]...)
+
+			// Success case
+			ref, err := tc.appendFunc(app, lset, 100, 30, tc.value)
+			require.NoError(t, err)
+			seriesMap[ref] = lset
+
+			if tc.expectedType == "histogram" {
+				histogramsMap[ref] = append(histogramsMap[ref], record.RefHistogramSample{Ref: chunks.HeadSeriesRef(ref), T: 30, H: &histogram.Histogram{}})
+				histogramsMap[ref] = append(histogramsMap[ref], record.RefHistogramSample{Ref: chunks.HeadSeriesRef(ref), T: 100, H: tc.value.(*histogram.Histogram)})
+			} else {
+				samplesMap[ref] = append(samplesMap[ref], record.RefSample{Ref: chunks.HeadSeriesRef(ref), T: 30, V: 0})
+				samplesMap[ref] = append(samplesMap[ref], record.RefSample{Ref: chunks.HeadSeriesRef(ref), T: 100, V: tc.value.(float64)})
+			}
+
+			// Error case: CT >= T
+			err = tc.appendErrorFunc(app, ref, lset, 100, 100, tc.value)
+			require.Error(t, err)
+			require.Equal(t, tc.expectedErrorMsg, err.Error())
+
+			require.NoError(t, app.Commit())
+
+			// Close the DB to ensure all data is flushed to the WAL
+			require.NoError(t, s.Close())
+
+			// Now read the WAL and verify its contents
+			sr, err := wlog.NewSegmentsReader(s.wal.Dir())
+			require.NoError(t, err)
+			defer sr.Close()
+
+			r := wlog.NewReader(sr)
+			dec := record.NewDecoder(labels.NewSymbolTable())
+
+			var (
+				allSeries  []record.RefSeries
+				samples    []record.RefSample
+				histograms []record.RefHistogramSample
+			)
+
+			for r.Next() {
+				rec := r.Record()
+				switch dec.Type(rec) {
+				case record.Series:
+					series, err := dec.Series(rec, nil)
+					require.NoError(t, err)
+					allSeries = append(allSeries, series...)
+				case record.Samples:
+					samples, err = dec.Samples(rec, samples[:0])
+					require.NoError(t, err)
+				case record.HistogramSamples:
+					histograms, err = dec.HistogramSamples(rec, histograms[:0])
+					require.NoError(t, err)
+				}
+			}
+
+			// Verify the contents
+			require.Len(t, allSeries, 1, "Expected 1 series")
+			if tc.expectedType == "histogram" {
+				require.Len(t, histograms, tc.expectedSampleCount, "Expected %d histograms", tc.expectedSampleCount)
+				require.Empty(t, samples, "Expected no float samples")
+			} else {
+				require.Len(t, samples, tc.expectedSampleCount, "Expected %d samples", tc.expectedSampleCount)
+				require.Empty(t, histograms, "Expected no histograms")
+			}
```

Comment:
This is awesome! Reading the WAL to verify that we appended the right thing is so much better than checking metrics! I think several tests in this file should be using this approach instead :) 

To make this usable for others as well, what do you think about extracting this logic into a separate function? Something like:

```go
func WalToChunkSample(t *testing.T, wal *wlog.WL) []chuncks.Sample {
 t.Helper()
 ...
}
```

Reply:
Done that. For now kept private to the package, later when people want to reuse it, they can export it if they want.